File: hetuflow/hetuflow-server/src/endpoint/api/v1/tasks.rs
use axum::{Json, extract::Path};
use fusion_common::model::IdUuidResult;
use fusion_web::{WebResult, ok_json};
use hetuflow_core::models::{SchedTask, TaskForCreate, TaskForQuery, TaskForUpdate};
use fusionsql::page::PageResult;
use serde_json::Value;
use utoipa_axum::router::OpenApiRouter;
use uuid::Uuid;

use crate::{application::ServerApplication, service::TaskSvc};

pub fn routes() -> OpenApiRouter<ServerApplication> {
  OpenApiRouter::new()
    .routes(utoipa_axum::routes!(query_tasks))
    .routes(utoipa_axum::routes!(create_task))
    .routes(utoipa_axum::routes!(get_task))
    .routes(utoipa_axum::routes!(update_task))
    .routes(utoipa_axum::routes!(delete_task))
    .routes(utoipa_axum::routes!(retry_task))
    .routes(utoipa_axum::routes!(cancel_task))
}

// parent: /api/v1/tasks
#[utoipa::path(
  post,
  path = "/query",
  request_body = TaskForQuery,
  responses(
    (status = 200, body = PageResult<SchedTask>)
  )
)]
async fn query_tasks(task_svc: TaskSvc, Json(input): Json<TaskForQuery>) -> WebResult<PageResult<SchedTask>> {
  let page_result = task_svc.page(input).await?;
  ok_json!(page_result)
}

#[utoipa::path(
  post,
  path = "/create",
  request_body = TaskForCreate,
  responses(
    (status = 200, description = "Success", body = IdUuidResult)
  )
)]
async fn create_task(task_svc: TaskSvc, Json(input): Json<TaskForCreate>) -> WebResult<IdUuidResult> {
  let id = task_svc.create_task(input).await?;
  ok_json!(id.into())
}

#[utoipa::path(
  get,
  path = "/{id}",
  params(
    ("id" = Uuid, Path, description = "Task ID")
  ),
  responses(
    (status = 200, description = "Success", body = Option<SchedTask>)
  )
)]
async fn get_task(task_svc: TaskSvc, Path(id): Path<Uuid>) -> WebResult<Option<SchedTask>> {
  let task = task_svc.get_by_id(id).await?;
  ok_json!(task)
}
#[utoipa::path(
  post,
  path = "/{id}/update",
  params(
    ("id" = Uuid, Path, description = "Task ID")
  ),
  request_body = TaskForUpdate,
  responses(
    (status = 200, description = "Success", body = Value)
  )
)]
async fn update_task(task_svc: TaskSvc, Path(id): Path<Uuid>, Json(input): Json<TaskForUpdate>) -> WebResult<Value> {
  task_svc.update_task(id, input).await?;
  ok_json!()
}
#[utoipa::path(
  delete,
  path = "/{id}",
  params(
    ("id" = Uuid, Path, description = "Task ID")
  ),
  responses(
    (status = 200, description = "Success", body = Value)
  )
)]
async fn delete_task(task_svc: TaskSvc, Path(id): Path<Uuid>) -> WebResult<Value> {
  task_svc.delete_task(id).await?;
  ok_json!()
}
#[utoipa::path(
  post,
  path = "/{id}/retry",
  params(
    ("id" = Uuid, Path, description = "Task ID")
  ),
  responses(
    (status = 200, description = "Success", body = Value)
  )
)]
async fn retry_task(task_svc: TaskSvc, Path(id): Path<Uuid>) -> WebResult<Value> {
  task_svc.retry_task(id).await?;
  ok_json!()
}
#[utoipa::path(
  post,
  path = "/{id}/cancel",
  params(
    ("id" = Uuid, Path, description = "Task ID")
  ),
  responses(
    (status = 200, description = "Success", body = Value)
  )
)]
async fn cancel_task(task_svc: TaskSvc, Path(id): Path<Uuid>) -> WebResult<Value> {
  task_svc.cancel_task(id).await?;
  ok_json!()
}

File: packages/hetuflow-sdk/src/api/tasks.ts
import { HetuflowClient } from '../utils/client.js';
import {
  TaskForCreate,
  TaskForUpdate,
  TaskForQuery,
  SchedTask,
  PageResult_SchedTask,
  IdUuidResult,
} from '../types/index.js';

export class TaskAPI {
  constructor(private client: HetuflowClient) {}

  /**
   * 创建 Task
   */
  async createTask(data: TaskForCreate): Promise<IdUuidResult> {
    return this.client.post<IdUuidResult>('/api/v1/tasks/create', data);
  }

  /**
   * 查询 Task 列表
   */
  async queryTasks(query: TaskForQuery): Promise<PageResult_SchedTask> {
    return this.client.post<PageResult_SchedTask>('/api/v1/tasks/query', query);
  }

  /**
   * 获取单个 Task
   */
  async getTask(id: string): Promise<SchedTask | null> {
    return this.client.get<SchedTask | null>(`/api/v1/tasks/${id}`);
  }

  /**
   * 更新 Task
   */
  async updateTask(id: string, data: TaskForUpdate): Promise<void> {
    return this.client.post<void>(`/api/v1/tasks/${id}/update`, data);
  }

  /**
   * 删除 Task
   */
  async deleteTask(id: string): Promise<void> {
    return this.client.delete<void>(`/api/v1/tasks/${id}`);
  }

  /**
   * 取消 Task
   */
  async cancelTask(id: string): Promise<void> {
    return this.client.post<void>(`/api/v1/tasks/${id}/cancel`);
  }

  /**
   * 重试 Task
   */
  async retryTask(id: string): Promise<void> {
    return this.client.post<void>(`/api/v1/tasks/${id}/retry`);
  }
}
